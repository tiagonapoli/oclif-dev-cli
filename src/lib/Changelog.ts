import * as findVersions from 'find-versions'
import { readFile, writeFile } from 'fs-extra'
import * as marked from 'marked'
import * as mdRenderer from 'marked-to-md'
import * as semver from 'semver'

const isVersionToken = (token: marked.Token): boolean => {
  return token.type === 'heading' && token.depth === 2
}

const semverize = (title: string) => {
  title = title.replace(/(\d+\.\d+)\.x/, '$1.99999')
  title = title.replace(/(\d+)\.x/, '$1.99999.99999')

  const versions = findVersions(title, { loose: true })
  if (!versions || versions.length === 0) {
    return null
  }

  const version = versions.sort(semver.compare).pop()
  return version
}

const getVersionFromTitle = (title: string) => {
  const version = semverize(title)
  if (!semver.valid(version)) {
    return null
  }

  return {
    version,
  }
}

export class Changelog {
  public static async getChangelog(path: string) {
    const changelog = new Changelog(path)
    await changelog.init()
    return changelog
  }

  private content: string
  private tokens: any
  constructor(private path: string) {}

  public async init() {
    this.content = await readFile(this.path, { encoding: 'utf8' })
    const lexer = new marked.Lexer()
    this.tokens = lexer.lex(this.content)
  }

  public flush() {
    return writeFile(this.path, this.content)
  }

  public getVersionContent(version: string) {
    const versionTokens = [] as marked.TokensList
    versionTokens.links = {}

    let started = false
    for (const token of this.tokens) {
      if (isVersionToken(token)) {
        if (started) {
          break
        }

        if (getVersionFromTitle(token.text)?.version === version) {
          started = true
        }
      }

      if (started) {
        versionTokens.push(token)
      }
    }

    if (!started) {
      throw new Error(`Version ${version} not found on Changelog`)
    }

    const versionContent = marked.parser(versionTokens, {
      renderer: mdRenderer(new marked.Renderer()),
    })

    return versionContent
  }
}
