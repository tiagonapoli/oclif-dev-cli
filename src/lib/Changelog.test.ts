import { join } from 'path'
import { Changelog } from './Changelog'

const mocksPath = join(__dirname, '__mocks__')

describe('Get version changes', () => {
  test.each([['0.0.1'], ['0.1.0-beta.0'], ['0.1.0']])('Version %s', async version => {
    const ch = await Changelog.getChangelog(join(mocksPath, 'changelog.md'))
    expect(ch.getVersionContent(version).split('\n')).toMatchSnapshot()
  })

  test(`Throws error when version doesn't exist`, async () => {
    const ch = await Changelog.getChangelog(join(mocksPath, 'changelog.md'))
    expect(() => ch.getVersionContent('0.0.0')).toThrow('Version 0.0.0 not found')
  })
})
