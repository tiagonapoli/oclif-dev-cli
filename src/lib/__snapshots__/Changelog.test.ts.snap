// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get version changes Version 0.0.1 1`] = `
Array [
  "## [0.0.1] - 2019-06-19",
  "",
  "- First release",
  "",
  "",
]
`;

exports[`Get version changes Version 0.1.0 1`] = `
Array [
  "## [0.1.0] - 2019-06-19",
  "### Added",
  "",
  "- Testing",
  "",
  "### Fixed",
  "",
  "- Fixing some bug",
  "",
  "",
]
`;

exports[`Get version changes Version 0.1.0-beta.0 1`] = `
Array [
  "## [0.1.0-beta.0] - 2019-06-19",
  "",
  "- First beta",
  "",
  "",
]
`;
